pkg/driver/controller_modify_volume.go

```diff
@@ -17,8 +20,149 @@ const (
+	}
+	if r.modifyDiskOptions.VolumeType != "" {
+		h.mergedRequest.modifyDiskOptions.VolumeType = r.modifyDiskOptions.VolumeType
+	}
+}
+
+// When a new request comes in, we look up requestHandlerMap using the volume ID of the request.
+// If there’s ModifyVolumeRequestHandler for the volume, meaning that there is inflight request(s) for the volume, we will first validate the
+// new request. If the new request is acceptable, we will send it to the Timer goroutine for the volume via a channel. In the Volume Timer goroutine,
+// the request will be merged with the existing request for the volume.
+// If there's no ModifyVolumeRequestHandler for the volume, meaning that there’s no inflight requests for the volume, we will create a
+// ModifyVolumeRequestHandler for the volume, add it to requestHandlerMap and start a Timer goroutine for the volume by calling startVolumeTimer
+// method. When the Volume Timer times out, it calls the ec2 API to perform the volume modification. The Timer goroutine sends back the results of
+// the ec2 API call to the CSI Driver main thread via errChan and resultChan.
+// Note that each volume with inflight requests has their own Timer goroutine which follows timeout schedule of their own.
+func (d *controllerService) addModifyVolumeRequest(ctx context.Context, volumeID string, r *modifyVolumeRequest) error {
+	v, ok := d.modifyVolumeManager.requestHandlerMap.Load(volumeID)
+	if ok {
+		requestHandler := v.(modifyVolumeRequestHandler)
+		requestHandler.mux.Lock()
+		err := requestHandler.validateModifyVolumeRequest(r)
+		if err == nil {
+			requestHandler.requestChan <- r
+		} else {
+			requestHandler.mux.Unlock()
+			return err
+		}
+	} else {
+		requestChan := make(chan *modifyVolumeRequest)
+		errChan := make(chan error)
+		resultChan := make(chan int64)
+		requestHandler := modifyVolumeRequestHandler{
+			requestChan:   requestChan,
+			errChan:       errChan,
+			resultChan:    resultChan,
+			mergedRequest: r,
+			volumeID:      volumeID,
+			mux:           &sync.Mutex{},
+		}
+		d.modifyVolumeManager.requestHandlerMap.Store(volumeID, requestHandler)
+
+		go d.startVolumeTimer(ctx, &requestHandler)
+	}
+	return nil
+}
+
+func (d *controllerService) startVolumeTimer(ctx context.Context, h *modifyVolumeRequestHandler) {
+	klog.V(4).InfoS("Starting ModifyVolume timer for ", "volumeID", h.volumeID)
+	for {
+		select {
+		case req := <-h.requestChan:
+			h.mergeModifyVolumeRequest(req)
+			// Unlock the mutex that was locked for validateModifyVolumeRequest call
+			// validateModifyVolumeRequest and mergeModifyVolumeRequest should happen in the same lock-context to prevent race
+			h.mux.Unlock()
+		case <-time.After(modifyVolumeRequestHandlerTimeout):
+			actualSizeGiB, err := d.executeModifyVolumeRequest(ctx, h.volumeID, h.mergedRequest)
+			if err != nil {
+				h.errChan <- err
+			} else {
+				h.resultChan <- actualSizeGiB
+			}
+			return
```

Comment:
Since we are only half a step away from that already, I suggest getting rid of the mux and relying on channels to handle concurrency: [Share Memory By Communicating](https://go.dev/blog/codelab-share).

The idea is to have a dedicated goroutine that processes _requests_ for a volume serially:

```
func (h *modifyVolumeRequestHandler) processRequests(d *controllerService, ctx context.Context) {
	for {
		select {
		case req := <-h.requestChan:
	        ...

		case <-time.After(modifyVolumeRequestHandlerTimeout):
		...

		case <-ctx.Done():
			return
		}
	}
}
```

Then, `addModifyVolumeRequest` can be simply reduced to:

```
func (d *controllerService) addModifyVolumeRequest(ctx context.Context, volumeID string, r *modifyVolumeRequest) error {
	handler, loaded := d.modifyVolumeManager.requestHandlerMap.LoadOrStore(volumeID, newModifyVolumeRequestHandler(volumeID))
	requestHandler := handler.(*modifyVolumeRequestHandler)
	if !loaded { 
		go requestHandler.processRequests(d, ctx)
	}
	requestHandler.requestChan 
	return nil
}
```